kernel-modules/probe/collector_probe.c

```diff
@@ -0,0 +1,280 @@
+  if ((flags & PF_KTHREAD) != 0) {
+    // we only want to process userspace threads.
+    return 0;
+  }
+
+  settings = get_bpf_settings();
+  if (settings == NULL) {
+    return 0;
+  }
+
+  if (!settings->capture_enabled) {
+    return 0;
+  }
+
+  call_filler(ctx, ctx, evt_type, settings, UF_NEVER_DROP);
+  return 0;
+}
+
+/**
+ * @brief Generic sys_enter_* program for any system call. It is responsible for
+ *        Verifying userspace settings and early processing of the syscall event.
+ *
+ *        The function will exit 0 (zero) regardless of outcome.
+ *
+ * @param id the syscall id
+ * @param ctx the context pointer as provided by the kernel
+ *
+ * @return 0 (regardless of outcomes)
+ */
+static __always_inline int enter_probe(long id, struct sys_enter_args* ctx) {
+  const struct syscall_evt_pair* sc_evt = NULL;
+  struct sysdig_bpf_settings* settings = NULL;
+  enum ppm_event_type evt_type = PPME_GENERIC_E;
+  int drop_flags = UF_ALWAYS_DROP;
+  struct sys_enter_args stack_ctx = {.id = id};
+
+  settings = get_bpf_settings();
+  if (settings == NULL) {
+    return 0;
+  }
+
+  if (!settings->capture_enabled) {
+    // capture_enabled is usually false when we have loaded the probe
+    // but not started capturing from a userspace perspective.
+    return 0;
+  }
+
+  sc_evt = get_syscall_info(id);
+  if (sc_evt == NULL) {
+    return 0;
+  }
+
+  evt_type = sc_evt->enter_event_type;
+  drop_flags = sc_evt->flags;
+
+  // To satisfy some verifier requiremnts in later parts of the falco plumbing/fillers,
+  // it is necessary to copy the context onto the stack.
+  memcpy(stack_ctx.args, _READ(ctx->args), sizeof(unsigned long) * NUM_SYS_ENTER_ARGS);
+
+  // stashing the args will copy it into a BPF map for later
+  // processing. This is a required step for the enter probe,
+  // and these args are subsequently pulled out of the map and
+  // written to the ring buffer.
+  //
+  // The args pointer must exist for the lifetime of this event, so we
+  // must not use _READ(stack_ctx.args) here.
+  if (stash_args(stack_ctx.args)) {
+    return 0;
+  }
+
+  // the fillers contain syscall specific processing logic, so we simply
+  // call into those and let the rest of falco deal with the event.
+  //
+  // It also handles the stack context problem, so we can pass both
+  // pointers through without issue.
+  call_filler(ctx, &stack_ctx, evt_type, settings, drop_flags);
+  return 0;
+}
+
+/**
+ * @brief Generic sys_exit_* program for any system call. It is responsible for
+ *        Verifying userspace settings and early processing of the syscall event.
+ *
+ *        The function will exit 0 (zero) regardless of outcome.
+ *
+ * @param id the syscall id
+ * @param ctx the eBPF context as provided by the kernel
+ *
+ * @return 0 (regardless of outcomes)
+ */
+static __always_inline int exit_probe(long id, struct sys_exit_args* ctx) {
+  const struct syscall_evt_pair* sc_evt = NULL;
+  struct sysdig_bpf_settings* settings = NULL;
```

Comment:
Just a heads up, `sysdig_bpf_settings` has been renamed in the latest changes of falco, along with a number of sysdig references, so we will need to be renaming them as well pretty soon.

Reply:
Good point! I think we also need more metadata sections defined at the bottom of this file because they're all validated at runtime by the falco userspace IIRC